sources/util/merge.js

```diff
@@ -0,0 +1,41 @@
+exports.mergePull = function (sources, next) {
+  if (!sources.length) return next()
+  let results = []
+  let pending = sources.length
+  sources.forEach(source => source.pull(onresult))
+  function onresult (result) {
+    if (result) results.push(result)
+    if (--pending === 0) onfinish()
+  }
+  function onfinish () {
+    if (!results.length) return next()
+    let messages = []
+    let finished = true
+    for (let result of results) {
+      if (result.messages) messages = messages.concat(result.messages)
```

Comment:
You can skip an array allocation on this hot path with this trick:
```suggestion
      if (result.messages) Array.prototype.push.apply(messages, result.messages)
```

Reply:
Nice, yes!